src/thermo/EEDFTwoTermApproximation.cpp

```diff
@@ -0,0 +1,699 @@
+    return 1./3. * m_gamma * simpson(f, x) / nDensity;
+}
+
+double EEDFTwoTermApproximation::electronMobility(const Eigen::VectorXd& f0)
+{
+    double nu = netProductionFreq(f0);
+    vector<double> y(options.m_points + 1, 0.0);
+    for (size_t i = 1; i < options.m_points; i++) {
+        // calculate df0 at i-1/2
+        double df0 = (f0(i) - f0(i-1)) / (m_gridCenter[i] - m_gridCenter[i-1]);
+        if (m_gridEdge[i] != 0.0) {
+            y[i] = m_gridEdge[i] * df0 /
+                   (m_totalCrossSectionEdge[i] + nu / m_gamma / pow(m_gridEdge[i], 0.5));
+        }
+    }
+    double nDensity = m_phase->molarDensity() * Avogadro;
+    auto f = ConstMappedVector(y.data(), y.size());
+    auto x = ConstMappedVector(m_gridEdge.data(), m_gridEdge.size());
+    return -1./3. * m_gamma * simpson(f, x) / nDensity;
+}
+
+void EEDFTwoTermApproximation::initSpeciesIndexCS()
+{
+    // set up target index
+    m_kTargets.resize(m_phase->nCollisions());
+    m_klocTargets.resize(m_phase->nCollisions());
+    m_inFactor.resize(m_phase->nCollisions());
+    for (size_t k = 0; k < m_phase->nCollisions(); k++) {
+        m_kTargets[k] = m_phase->targetIndex(k);
+        // Check if it is a new target or not :
+        auto it = find(m_k_lg_Targets.begin(), m_k_lg_Targets.end(), m_kTargets[k]);
+
+        if (it == m_k_lg_Targets.end()){
+            m_k_lg_Targets.push_back(m_kTargets[k]);
+            m_klocTargets[k] = m_k_lg_Targets.size() - 1;
+        } else {
+            m_klocTargets[k] = distance(m_k_lg_Targets.begin(), it);
+        }
+
+        const auto& kind = m_phase->collisionRate(k)->kind();
+
+        if (kind == "ionization") {
+            m_inFactor[k] = 2;
+        } else if (kind == "attachment") {
+            m_inFactor[k] = 0;
+        } else {
+            m_inFactor[k] = 1;
+        }
+    }
+
+    m_X_targets.resize(m_k_lg_Targets.size());
+    m_X_targets_prev.resize(m_k_lg_Targets.size());
+    for (size_t k = 0; k < m_X_targets.size(); k++) {
+        size_t k_glob = m_k_lg_Targets[k];
+        m_X_targets[k] = m_phase->moleFraction(k_glob);
+        m_X_targets_prev[k] = m_phase->moleFraction(k_glob);
+    }
+
+    // set up indices of species which has no cross-section data
+    for (size_t k = 0; k < m_phase->nSpecies(); k++) {
+        auto it = std::find(m_kTargets.begin(), m_kTargets.end(), k);
+        if (it == m_kTargets.end()) {
+            m_kOthers.push_back(k);
+        }
+    }
+}
+
+void EEDFTwoTermApproximation::checkSpeciesNoCrossSection()
+{
+    // warn that a specific species needs cross-section data.
+    for (size_t k : m_kOthers) {
+        if (m_phase->moleFraction(k) > options.m_moleFractionThreshold) {
+            writelog("EEDFTwoTermApproximation:checkSpeciesNoCrossSection\n");
+            writelog("Warning:The mole fraction of species {} is more than 0.01 (X = {:.3g}) but it has no cross-section data\n", m_phase->speciesName(k), m_phase->moleFraction(k));
+        }
```

Comment:
All the calls to `writelog` in this file should probably be updated to be versions of `warn_user`, or possibly just to throw `CanteraError` exceptions. Also, this line is significantly beyond our normal limit of 88 characters.